Addition of DCGMDevicePool for NVIDIA Profiling Metrics

lhlawson opened the pull request:
Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

Addition of a DCGM Device pool that connects to a local standalone nv-hostengine in order to query NVIDIA Profiling metrics.  This PR is just the device pool and is not currently used by any classes in the code base.  This is a prerequisite for PR #1914.

Details of change:
- Added DCGMDevicePool class and throw.
- The user is expected to have setup a local instance of nv-hostengine for the device pool to connect to.  

Related Features:
- Relates to #1902
- Fixes #1915
- Fixes #1904

Long term it may be worth considering adding a way for the user to specify an IP address for the nv-hostengine.

lhlawson committed:
Added DCGMDevicePool

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

dannosliwcd commented on service/src/DCGMDevicePool.cpp:
```diff
@@ -0,0 +1,160 @@
+        result = dcgmGetAllSupportedDevices(m_dcgm_handle, dcgm_dev_id_list, &m_dcgm_dev_count);
+        check_dcgm_result(result, "Error fetching DCGM supported devices.");
+
+        for(int i = 0; i < m_dcgm_dev_count; ++i) {
```
There's a missing space at `for (`. Also some inconsistent spacing after `//` style comments, and I think this project uses pre-c++11-style nested templates (space between closing `> >`).

You can let a script handle formatting issues. I have a `clang-format` configuration file at  https://gist.github.com/dannosliwcd/4b0c593968820438889c6461395c0eca that I've been tweaking whenever I receive code style suggestions in code review.

dannosliwcd commented on service/src/DCGMDevicePoolImp.hpp:
```diff
@@ -0,0 +1,75 @@
+
+            unsigned short m_dcgm_field_ids[M_FIELD_IDS];
+
+            std::vector<std::vector<dcgmFieldValue_v1>> m_dcgm_field_values;
```
Maybe add a comment about how this is indexed. I think the inner one is obviously indexed by value, but the outer vector's index isn't as clear without looking where it is already used.

dannosliwcd commented on service/src/DCGMDevicePool.cpp:
```diff
@@ -0,0 +1,160 @@
+
+        for(int i = 0; i < m_dcgm_dev_count; ++i) {
+            std::vector<dcgmFieldValue_v1> field_values(sizeof(m_dcgm_field_ids)/sizeof(m_dcgm_field_ids[0]));
+            m_dcgm_field_values.push_back(field_values);;
```
Double `;;` here. Might get handled by clang-format but I'm not sure.

dannosliwcd commented on service/src/DCGMDevicePool.cpp:
```diff
@@ -0,0 +1,160 @@
+
+        //Setup Field Group
+        result = dcgmFieldGroupCreate(m_dcgm_handle, sizeof(m_dcgm_field_ids)/sizeof(m_dcgm_field_ids[0]), &m_dcgm_field_ids[0],
+                                      (char *)"geopm_fields", &m_field_group_id);
```
`(char *)"geopm_fields"` seems risky, toward hitting undefined behavior. DCGM docs say it's input only, but I'm not sure why they don't have it as const. Might be safer to put that string in a local array and send that array's address to this function.

dannosliwcd commented on service/src/DCGMDevicePool.cpp:
```diff
@@ -0,0 +1,160 @@
+    DCGMDevicePoolImp::~DCGMDevicePoolImp()
+    {
+        //Shutdown DCGM
+        dcgmStatusDestroy(NULL);
```
Does the field group also need to be destroyed?

lhlawson replied:
Yes!

dannosliwcd commented on service/src/DCGMDevicePool.cpp:
```diff
@@ -0,0 +1,160 @@
+        if (result != DCGM_ST_OK) {
+            throw Exception("DCGMDevicePool::" + std::string(__func__) + ": "
+                            + error + ": " + errorString(result),
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
```
All errors will show `check_dcgm_result`, line 109. Might pass through the function name and line number as arguments to this function.

dannosliwcd commented on service/src/DCGMDevicePool.cpp:
```diff
@@ -0,0 +1,160 @@
+    }
+
+    double DCGMDevicePoolImp::field_value(int accel_idx, int geopm_field_id) const {
+        double result;
```
Maybe throw an exception in debug builds if type is not DCGM_FT_DOUBLE

dannosliwcd replied:
Maybe also a not-just-debug check that all fields are double at initialization time, if that's possible.

lhlawson replied:
Moved to doing prechecking of all fields at init time.  Removed the PCIE_* fields as they are NOT double formatted.  We can re-add them if needed and add additional checking for the appropriate data type.

lhlawson committed:
Addressed Review Feedback & Added retry case on field create

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

lhlawson commented:
A few updates as part of the review process:
- Updated initialization to handle the case where a field group needs to be destroyed and recreated
- Updated initialization to verify all FIELD IDS are of types we handle.  
- Removed PCIE metrics

cmcantalupo commented on service/src/DCGMDevicePool.hpp:
```diff
@@ -0,0 +1,61 @@
+            DCGMDevicePool() = default;
+            virtual ~DCGMDevicePool() = default;
+
+            virtual int dcgm_device() const = 0;
+            virtual double field_value(int accel_idx, int geopm_field_id) const = 0;
+            virtual void update_field_value(int accel_idx) = 0;
+            virtual void field_update_rate(int accel_idx, int field_update_rate) = 0;
+            virtual void max_storage_time(int accel_idx, int max_storage_time) = 0;
+            virtual void max_samples(int accel_idx, int max_samples) = 0;
```
Please add doxygen comments.

cmcantalupo commented on service/src/DCGMDevicePool.hpp:
```diff
@@ -0,0 +1,61 @@
+    {
+        public:
+            enum geopm_dcgm_field_ids_e {
+                M_SM_ACTIVE,
+                M_SM_OCCUPANCY,
+                M_DRAM_ACTIVE,
+                M_FIELD_IDS
```
Add doxygen

cmcantalupo commented on service/src/DCGMDevicePool.hpp:
```diff
@@ -0,0 +1,61 @@
+
+namespace geopm
+{
+    class DCGMDevicePool
```
Add doxygen

lhlawson committed:
Fixed mis-rebase/merge that restuled in extraneous inputs to DCGM controls.  Added Doxygen comments

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

dannosliwcd commented on service/src/DCGMDevicePool.cpp:
```diff
@@ -0,0 +1,186 @@
+    {
+        //Shutdown DCGM
+        dcgmStatusDestroy(NULL);
+        dcgmFieldGroupDestroy(m_dcgm_handle, m_field_group_id);
```
Github is not letting me comment on line 127. This comment goes with line 127.

`dcgmStatusDestroy` accepts a `dcgmStatus_t` argument, rather than a pointer. Also, there is no call to `dcgmStatusCreate`.

dannosliwcd commented on service/src/DCGMDevicePool.cpp:
```diff
@@ -0,0 +1,186 @@
+
+        // Note: Currently we are using dcgmWatchFields, but may transition to using
+        //       dcgmProfWatchFields at a later date
+        result = dcgmWatchFields(m_dcgm_handle, DCGM_GROUP_ALL_GPUS, m_field_group_id, m_update_freq,
+                                 m_max_keep_age, m_max_keep_sample);
```
If I understand the docs, I think this is requesting that DCGM cache up to m_max_keep_sample (100) most-recent values, internally sampled every 1000 usec, keeping up to m_max_keep_age (1) seconds (i.e., up to 1000, were it not for the explicit sample limit).

Does this mean that each call to `dcgmGetLatestValuesForFields` is ignoring 99 samples from each call to DCGM? Could we alternatively get sub-geopm-sample-rate information by using `dcgmGetValuesSince_v2` that gets an array of dcgm samples that were observed between geopm requests? This could be useful for observing spikes in non-monotonically-counting signals, like the utilization signals. In that case, we could report the average or peak utilization between geopm samples instead of the point-in-time utilization when geopm gets its samples.

The docs aren't clear if that's the difference between those functions in their long-form descriptions. This is all just assumptions based on their names.

lhlawson replied:
In testing using dcgmGetValuesSince wasn't necessary after increasing default polling rate to ever 100ms.  Some experimental agents currently use a lower polling rate, but can use the controls provided through an upcoming IOGroup #1914 to set the rate as needed.

dannosliwcd commented on service/src/DCGMDevicePool.cpp:
```diff
@@ -0,0 +1,186 @@
+
+    double DCGMDevicePoolImp::field_value(int accel_idx, int geopm_field_id) const {
+        double result;
+        result = m_dcgm_field_values.at(accel_idx).at(geopm_field_id).value.dbl;
```
Can also return NAN if .status is not DCGM_ST_OK. Hopefully doesn't happen now that we have the lower update rate, but might be good in case anyone changes the rate.

lhlawson committed:
Addressed review feedback: default query rate, naming, doxygen updates

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

lhlawson commented on service/src/DCGMDevicePool.hpp:
```diff
@@ -0,0 +1,99 @@
+            /// @brief Query DCGM for the latest value for an accelerator.
+            ///        Note that this is the last value DCGM cahced.  This
+            //         updates the DCGM device pool stored value that is provided
+            //         via the field_value function
```
```suggestion
            //         via the sample_field_value function
```

dannosliwcd reviewed (dismissed):
No additional comments from me. I've been using a slightly-stale version of this for about a week, and it seems to be working as expected since the change to a 100ms sample period.

lhlawson committed:
Fixed re-appearing issue with misplaced ; and char pointer vs array

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

cmcantalupo commented on service/src/DCGMDevicePool.hpp:
```diff
@@ -0,0 +1,99 @@
+    class DCGMDevicePool
+    {
+        public:
+            enum geopm_dcgm_field_ids_e {
+                /*!
+                 * @brief Field ID associated with DCGM SM Active metrics
+                 */
+                M_SM_ACTIVE,
+                /*!
+                 * @brief Field ID associated with SM Occupancy metrics
+                 */
+                M_SM_OCCUPANCY,
+                /*!
+                 * @brief Field ID associated with DCGM DRAM Active metrics
+                 */
+                M_DRAM_ACTIVE,
+                /*!
+                 * @brief Number of valid field ids
+                 */
+                M_FIELD_IDS
+            };
```
```suggestion
            enum m_field_id_e {
                /*!
                 * @brief Field ID associated with DCGM SM Active metrics
                 */
                M_FIELD_ID_SM_ACTIVE,
                /*!
                 * @brief Field ID associated with SM Occupancy metrics
                 */
                M_FIELD_ID_SM_OCCUPANCY,
                /*!
                 * @brief Field ID associated with DCGM DRAM Active metrics
                 */
                M_FIELD_ID_DRAM_ACTIVE,
                /*!
                 * @brief Number of valid field ids
                 */
                M_NUM_FIELD_ID
            };
```

cmcantalupo commented on service/src/DCGMDevicePool.hpp:
```diff
@@ -0,0 +1,99 @@
+            /// @param [in] accel_idx The index indicating a particular
+            ///        accelerator.
+            /// @return The value for the specified field
+            virtual double sample_field_value(int accel_idx, int geopm_field_id) const = 0;
```
```suggestion
            virtual double sample(int accel_idx, int field_id) const = 0;
```

cmcantalupo commented on service/src/DCGMDevicePool.hpp:
```diff
@@ -0,0 +1,99 @@
+
+            /// @brief Number of accelerators that support DCGM on the platform.
+            /// @return Number of accelerators.
+            virtual int dcgm_device() const = 0;
```
```suggestion
            virtual int num_device() const = 0;
```

cmcantalupo commented on service/src/DCGMDevicePool.hpp:
```diff
@@ -0,0 +1,99 @@
+            //         via the sample_field_value function
+            /// @param [in] accel_idx The index indicating a particular
+            ///        accelerator.
+            virtual void update_field_value(int accel_idx) = 0;
```
update_field_value -> update?

cmcantalupo commented on service/src/DCGMDevicePool.hpp:
```diff
@@ -0,0 +1,99 @@
+            virtual void update_field_value(int accel_idx) = 0;
+            /// @brief Set field update rate for DCGM devices.
+            /// @param [in] field_update_rate DCGM update rate in microseconds.
+            virtual void field_update_rate(int field_update_rate) = 0;
```
field_update_rate -> update_rate?  What is this rate determining?  I'm really not sure what this method does.

cmcantalupo commented on service/src/DCGMDevicePool.hpp:
```diff
@@ -0,0 +1,99 @@
+            virtual void field_update_rate(int field_update_rate) = 0;
+            /// @brief Set maximum storage time for for DCGM devices.
+            /// @param [in] max_storage_time maximum storage time in seconds
+            virtual void max_storage_time(int max_storage_time) = 0;
```
I don't understand what "storage time" means, please define

lhlawson replied:
Storage time is defined by the [DCGM API](https://docs.nvidia.com/datacenter/dcgm/1.6/dcgm-api/group__DCGMAPI__FI.html#group__DCGMAPI__FI_1gd04fa06405af6165fd85310dee9eae2a), it's how long the DCGM process will store those samples before dropping them.

cmcantalupo commented on service/src/DCGMDevicePool.hpp:
```diff
@@ -0,0 +1,99 @@
+            /// @param [in] max_storage_time maximum storage time in seconds
+            virtual void max_storage_time(int max_storage_time) = 0;
+            /// @brief Set maximum samples to store for for DCGM devices.
+            /// @param [in] max_samples maximun number of samples to store
```
I'm not sure what this method does either

cmcantalupo reviewed (commented):
I have made some suggestions to the DCGMDevicePool header.  I'm sure these will impact the implementation in this PR and also the implementationf or the DGCMIOGroup.  Please address.  I would like better documentation for the APIs before reviewing the implementation.

cmcantalupo commented on service/src/DCGMDevicePool.cpp:
```diff
@@ -0,0 +1,188 @@
+        check_dcgm_result(result, "Error fetching DCGM supported devices.", __LINE__);
+
+        for (int i = 0; i < m_dcgm_dev_count; ++i) {
+            std::vector<dcgmFieldValue_v1> field_values(sizeof(m_dcgm_field_ids)/sizeof(m_dcgm_field_ids[0]));
```
```suggestion
            std::vector<dcgmFieldValue_v1> field_values(M_NUM_FIELD_ID);
```

cmcantalupo commented on service/src/DCGMDevicePool.cpp:
```diff
@@ -0,0 +1,188 @@
+        //Setup Field Group
+
+        char geopm_group[] = "geopm_field_group";
+        result = dcgmFieldGroupCreate(m_dcgm_handle, sizeof(m_dcgm_field_ids)/sizeof(m_dcgm_field_ids[0]), &m_dcgm_field_ids[0],
```
please get rid of all of these sizeof(...) / sizeof(...)

cmcantalupo committed:
Clean up attempt for DCGMDevicePool.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

lhlawson committed:
Second pass of DCGMDevicePool cleanup attempt

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

lhlawson committed:
DCGMDevPool documentation updates

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

cmcantalupo reviewed (approved):
LGTM
